### PR TITLE
Fix extra `useQuery` result frames

### PR DESCRIPTION
**Pull request by benjamn**
This PR builds on #9508, fixing the unexpected frame by transitioning to a `loading: true` result synchronously when variables change or the query is refetched.

Since we do not want to commit to making any network requests synchronously during rendering, I have come up with a new way to block all network requests, by passing a `fetchBlockingPromise: Promise<boolean>` option that intercepts a part of the request pipeline that is already safely asynchronous.

As a happy side effect, this `fetchBlockingPromise` technique should also prevent making any network requests for extra `ObservableQuery` objects created during development when using React's `<StrictMode>`, since the extra component renders are never mounted, and thus `useEffect` never fires, leaving the `fetchBlockingPromise` unresolved until it times out and the fetch gets silently canceled (as desired).

### CODE REVIEW

**Comment by dylanwulf** [comment]
I have confirmed that this fixes #9549 and it might also fix #9145 and #9580

File: src/core/watchQueryOptions.ts [hunk 1]
```diff
@@ -145,6 +145,14 @@ export interface WatchQueryOptions<TVariables = OperationVariables, TData = any>
...
    */
   refetchWritePolicy?: RefetchWritePolicy;
+
+  /**
+   * If provided, stalls any network activity for this request until the Promise
+   * has resolved. If the Promise resolves to true, the network request will
+   * proceed. If the Promise resolves to false, the network request will be
+   * silently discarded.
+   */
+  fetchBlockingPromise?: Promise<boolean>;
```
Comment (brainkim) [inline src/core/watchQueryOptions.ts:155]: This is a new API for React super-strict mode. It is technically a public API, but no one should be using this when they are using the ObservableQuery API directly.

TL;DR A sludge for React 18.

File: src/core/QueryManager.ts [hunk 2]
```diff
@@ -1390,13 +1391,25 @@ export class QueryManager<TStore> {
...
+          fetchPolicy,
+          errorPolicy,
+        },
+      );
+
+      // If we have a fetchBlockingPromise, wait for it to be resolved before
+      // allowing any network requests, and only proceed if fetchBlockingPromise
+      // resolves to true. If it resolves to false, the request is discarded.
+      return fetchBlockingPromise ? fetchBlockingPromise.then(
+        ok => ok ? get() : Observable.of<ApolloQueryResult<TData>>()
```
Comment (brainkim) [inline src/core/QueryManager.ts]: nit: whatever I like the name I guess

File: src/core/QueryManager.ts [hunk 3]
```diff
@@ -1390,13 +1391,25 @@ export class QueryManager<TStore> {
...
+      ) : get();
```
Comment (brainkim) [inline src/core/QueryManager.ts:1427]: This is technically .
.
z̢̛͈̱̯͎̬̒̎͒̒̑̒́̃̎a̴͕̜͍͈̳̽̉͋̋̄̈́͞͡l̷̗͖̝̖̜̻̹͐̋͋͒̂̚͘̚͞ǵ̵̠̤̖̠̫͇̭̝̬͐̈̔̀͌͘͢͠͝ò̵͎̪̖͇̙̪̜̖͆̎̉͑̅̎̅͜
̧̛̛̩̯̣͔̥̻̿͌̅͛̑͘͜͝͡
.
.

**Comment by brainkim** [comment]
I think this could be merged for the sake of the release candidate but we should definitely loop around on the `fetchBlockingPromise` API because it is technically public.

A surgical strike by @benjamn.

cc @hwillson @jpvajda

**Comment by yusufumac** [comment]
@benjamn I came across to bug using fetchMore function which is indirectly caused by fetchBlockingPromise. Because resultsFromLink  ([source](https://github.com/apollographql/apollo-client/blob/85e8fc3bd198ab04e0c16c3a27054425cbc3f905/src/core/QueryManager.ts#L1409)) can't execute query - by your design-, fetchQuery function returns undefined in fetchMore causes exception [here](https://github.com/apollographql/apollo-client/blob/85e8fc3bd198ab04e0c16c3a27054425cbc3f905/src/core/ObservableQuery.ts#L454).

**Comment by benjamn** [comment]
@yusufumac Thank you! Would you mind opening a new issue, ideally with a small [reproduction](https://github.com/apollographql/react-apollo-error-template/)? I have some guesses about what the problem might be, but it would be great to see some concrete code. 

Are you saying `resultsFromLink` remains permanently blocked when you use `fetchMore`? Is it possible the component in question is never actually mounted? I ask because `<React.StrictMode>` generates double renderings and discards half of them without running `useEffect` callbacks, which could potentially be a source of components that never mount, so those `fetchBlockingPromise`s resolve to `false` and stay that way permanently (or until `fetchBlockingPromise` is removed from the options). Does any of that sound familiar/plausible from your debugging?

**Comment by benjamn** [comment]
@yusufumac An update relevant to your concerns: PR #9636 removes the `fetchBlockingPromise` logic, and was just released in `@apollo/client@3.6.1` (still tagged as `next` rather than `latest` on npm for now). Please let us know if you have a chance to try `npm i @apollo/client@next` in your reproduction!

**Comment by yusufumac** [comment]
@benjamn Thanks for the update. It was a busy week, so I couldn't create a reproducible case for you. I downgraded to 3.5.10 and didn't have the same issue. I will try 3.6.1 and will let you know if the issue still there.
